File: lib/skill_search_skill.dart
import 'dart:typed_data';

/// A skill entry in the skill search overlay.
class SkillSearchSkill {
  /// Unique skill identifier.
  final String id;

  /// Display name shown in the row.
  final String name;

  /// SF Symbol name, used as fallback when [iconBytes] is `null`.
  final String? icon;

  /// Keyboard shortcut display (e.g. `"⌘Q"`).
  final String shortcut;

  /// PNG image bytes. Preferred over [icon] when present.
  final Uint8List? iconBytes;

  const SkillSearchSkill({
    required this.id,
    required this.name,
    this.icon,
    required this.shortcut,
    this.iconBytes,
  });

  Map<String, dynamic> toMap() => {
    'id': id,
    'name': name,
    'shortcut': shortcut,
    if (icon != null) 'icon': icon,
    if (iconBytes != null) 'iconBytes': iconBytes,
  };
}

File: macos/shadow_action_skill/Sources/shadow_action_skill/SkillSearch/SkillSearchSkill.swift
import Foundation

// MARK: - SkillSearchSkill

/// A single skill entry in the SkillSearch popup.
///
/// Decoded from the Flutter JSON payload. Immutable value type. (Predictability)
struct SkillSearchSkill: Identifiable, Equatable {
    /// Unique skill identifier (e.g. "dictation", "screenshot")
    let id: String
    /// Display name shown in the row (e.g. "Polish my write")
    let name: String
    /// SF Symbol name for fallback display (e.g. "mic.fill", "camera.fill")
    let icon: String?
    /// Keyboard shortcut for display (e.g. "⌘Q") — single combined string from Flutter
    let shortcut: String
    /// PNG image bytes rendered by Flutter (preferred over SF Symbol when present)
    let iconBytes: Data?
}

// MARK: - Flutter Parsing

extension SkillSearchSkill {
    /// Parse an array of skills from the Flutter JSON arguments.
    ///
    /// Expected format: `{"skills": [{"id": "...", "name": "...", "icon": "...", "shortcut": "⌘Q", "iconBytes": Data}, ...]}`
    /// `icon` and `iconBytes` are both optional — at least one should be present.
    /// The router converts `FlutterStandardTypedData` → `Data` before this is called.
    /// Returns `nil` if the payload is malformed.
    static func fromFlutterArguments(_ arguments: Any?) -> [SkillSearchSkill]? {
        guard let dict = arguments as? [String: Any],
              let skillsArray = dict["skills"] as? [[String: Any]] else {
            return nil
        }
        return skillsArray.compactMap { item in
            guard let id = item["id"] as? String,
                  let name = item["name"] as? String,
                  let shortcut = item["shortcut"] as? String else {
                return nil
            }
            let icon = item["icon"] as? String
            let iconBytes = item["iconBytes"] as? Data
            return SkillSearchSkill(id: id, name: name, icon: icon, shortcut: shortcut, iconBytes: iconBytes)
        }
    }
}
